### PR TITLE
442 commitverification reduce undue penalties on students submitting early

### LINKED ISSUE #442: CommitVerification: Reduce Undue Penalties on Students Submitting Early
## Overview
Students passing off early face undue, unnecessary-positive Commit Verification warnings if they haven't committed on enough days.

**Give students credit towards their "required days with commits" for each day early that they pass off.**

## Discussion
The main point of the "day" requirement of the commit verification system is to ensure that students are not procrastinating work and doing it all at once.

We already have an informal policy that students passing off early are not subject to this requirement since they have clearly already worked ahead. The thinking goes that, if they have already completed the project now— before it's due—, then they could easily add an empty commit tomorrow and another empty commit the day after and submit with no issues.

However, rather than rewarding students for working ahead, we end up inadvertently penalizing these students if they complete the phase early, but in one day.

## Proposal
When evaluating the `git` minimum day requirement, we perform the following algorithm (notice the new steps 2, 3, and 4):
1. Do the existing check (`days_with_commits >= minimum_days_with_commits_phase_threshold`)
2. If they pass this check, then they pass the minimum days requirement
3. Count the number of full days between their most recent commit and the phase deadline (call this `early_days`)
4. If `early_days + days_with_commits >= minimum_days_with_commits_phase_threshold`, then pass the requirement.
    * Consider also displaying a notification explaining that we gave them grace on the commit penalty because they finished the project early
5. Otherwise, fail the requirement as normal

### Technical Notes
1. This will require giving some additional context to the outermost layer of the commit verification system since it currently has no idea when the assignment is due. It is already divided into layers of responsibility, and the inner layers will not need to be modified.
2. This would be a good opportunity to further split the responsibilities of the verification system internal components so that the exact error conditions and message strings could be dynamically loaded and provided with a pre-defined context object.
3. We could implement the "warning notification" with the new functionality being introduced in #422.
4. All things considered, this should result in only a minor modification to the condition evaluation code.

## Instructor Feedback
@leesjensen @jerodw
1. Is this proposal consistent with your view and vision of the course? 
5. Does it faithfully extend the requirements originally proposed in #160?

### PR DESCRIPTION
**Pull request by frozenfrank**
## Overview

Gives students credit towards their "required days with commits" for each day early that they pass off. Resolves #442.

## Details

- While some underlying dependencies were updated and test cases added, the actual change here has been minimized
    - Key change is commit f404546e
    - Key supporting change is commit 90eb7ee0
    - Key performance change is commit 9a082f87
    - Key organizational change is commit db075c63
- Changes the error condition for days with commits:
    - Condition: `daysWithCommits + daysSubmittedEarly < requiredDaysWithCommits`
    - Error: `"Did not commit on enough days to pass off (%d/%d)."` (no change)
- Adds a warning condition when submitting early helped bypass the requirements:
    - Condition: `!insufficientDaysWithCommits && daysWithCommits < requiredCommits && daysSubmittedEarly > 0`
    - Error: `"Committed %d of %d required days, but early completion made up the difference."`
- Adds tests for new method `LateDayCalculator::getNumDaysEarly()` to ensure math calculates properly
- Adjusts dependencies on `LateDayCalculator` and `Scorer` to:
    - obey Open-Closed principle (see #508)
    - preserve a single instance for an entire grading session
- Introduces result caching in `LateDayCalculator` to avoid repetitive round-trips to databases and Canvas APIs that retrieve necessary context for grading
- Read the commit messages for a more detailed list of changes

## Testing

- [x] Added/updated unit tests
- [x] Tested edge cases
- [ ] Manual testing (if needed)

While I added unit tests against `LateDayCalculator::getNumDaysEarly()`, I did not add overall integration tests to ensure the entire commit verification response occurs properly. If you need, I'll do that in this PR, or I can convert that to a separate issue.

## Additional Notes

Test case for `LateDayCalculator::getNumDaysEarly()`:

days-early

### CODE REVIEW

**Comment by frozenfrank** [comment]
The regressions tests originally failed because the new code requires context which was not needed before. Since the context isn't _really_ required, I mocked out that particular dependency so the tests can continue to focus on just the things they are explicitly testing.

**Comment by frozenfrank** [comment]
Update: I'm going to add a `git` integration test to specifically cover this new case. Please still review the PR since the changes from here will be minimal.

**Comment by webecke** [comment]
> Introduces result caching in LateDayCalculator to avoid repetitive round-trips to databases and Canvas APIs that retrieve necessary context for grading

This might cause issues. How long does the cache live, because soon holidays could change without a restart. Additionally, not every student has the same due date, some students have professor extensions which are dealt with in Canvas.

**Review by webecke — COMMENTED** [review]
I'm not officially requesting changes, but I am curious to get more info on the caching. The early turn in stuff LGTM

**Review by 19mdavenport — APPROVED** [review]

**Review by webecke — APPROVED** [review]
LGTM!

File: src/main/java/edu/byu/cs/autograder/score/LateDayCalculator.java [hunk 1]
```diff
@@ -12,52 +12,111 @@
...
+ * The calculation of the late/early days requires additional information which must be requested from a network.
+ * This information is fetched and cached so that subsequent requests requiring the details
+ * do not create a performance bottleneck.
+ * <br>
+ * It is intended that an instance of this class will be shared and reused for at least an entire grading session.
+ * Since the simple cache preserves all metadata, it would be efficient to restart the object for each new grading
+ * session to avoid a memory leak.
+ * <br>
+ * CONSIDER: Implementing a max-cache-count cache which would not require a new instance in order to avoid memory leaks.
+ * At the present time, this system is more complex than is necessary.
```
[the remark this answers is not in the document]
Comment (frozenfrank) [reply src/main/java/edu/byu/cs/autograder/score/LateDayCalculator.java:35]: Good point. Allow me to clarify the written words here because are currently unclear and contradictory.

File: src/test/java/edu/byu/cs/autograder/git/GitHelperTest.java [hunk 2]
```diff
@@ -267,4 +267,19 @@ void passoffMissingTailHash() {
                     utils.generalCommitVerificationResult(false, 2, 2, true))
         ));
     }
+
+    @Test
+    void passoffEarlyInsufficientDays() {
+        utils.setGradingContext(utils.generateGradingContext(1, 3, 10,  1));
```
Comment (frozenfrank) [inline src/test/java/edu/byu/cs/autograder/git/GitHelperTest.java:273]: White space

File: src/test/java/edu/byu/cs/autograder/git/GitHelperUtils.java [hunk 3]
```diff
@@ -237,7 +244,10 @@ GitEvaluator<CommitVerificationResult> evaluateRepo(@Nullable CommitThreshold mi
         return evaluateRepo(gradingContext, minThreshold);
     }
     GitEvaluator<CommitVerificationResult> evaluateRepo(GradingContext gradingContext, @Nullable CommitThreshold minThreshold) {
-        return evaluateRepo(new GitHelper(gradingContext), minThreshold);
+        LateDayCalculator lateDayCalculator = new MockLateDayCalculator(submitDaysEarly);
+        return evaluateRepo(
+                new GitHelper(gradingContext, new DefaultGitVerificationStrategy(lateDayCalculator)),
```
Comment (frozenfrank) [inline src/test/java/edu/byu/cs/autograder/git/GitHelperUtils.java:249]: I'll unwrap this line more

File: src/test/java/edu/byu/cs/autograder/score/LateDayCalculatorTest.java [hunk 4]
```diff
@@ -77,146 +78,222 @@ void getNumDaysLateWithoutHolidays() {
...
+
+    @Test
+    void getNumDaysEarly() {
+        // Initialize without holidays
+        LateDayCalculator lateDayCalculator = new LateDayCalculator();
+        lateDayCalculator.initializePublicHolidays();
+
+        // See images: days-early
+        String dueDateStr = "1999-11-19 11:59:00 PM -07:00";
+        ExpectedDaysDiff[] expectedDaysLate = {
```
Comment (frozenfrank) [inline src/test/java/edu/byu/cs/autograder/score/LateDayCalculatorTest.java:219]: Should refer to days early, not late

**Review by frozenfrank — COMMENTED** [review]
I have a few clean up tasks to do, but otherwise the functionality is done
